batches: partially revert making `batch_spec_id` nullable on batch change

courier-new opened the pull request:
This PR reverts most of https://github.com/sourcegraph/sourcegraph/pull/44840, except for:
- The changes related to the batch change name pattern constraint
- The original migrations which make `batch_spec_id` nullable

Since the migration has already been run, we revert the migration by leaving the original one completely intact, then adding a new one which runs the original migration's `down` as its `up`. None of this has been released to customers yet, so the two migrations should just run one after another when customers go to upgrade to 4.4, leaving the DB end state as if nothing happened.

We're choosing to revert the commit because fixing one regression from it (https://github.com/sourcegraph/sourcegraph/pull/46237) actually revealed there were more problems with the frontend handling the changes. The frontend is unfortunately architected in such a way right now that bakes in the assumption that if we have a batch change, there will also always be a batch spec. TS gave us false confidence about working around that assumption because we don't have the TS no unchecked index access flag enabled on the compiler, but in reality we're looking at a sizable refactor to get around all the queries and mutations that now conditionally need to be called only when there's a batch spec, and the context that needs to be split up to allow working with the batch change and batch spec in isolation.

## Test plan

Tests pass, verified no null exceptions when viewing draft batch changes from the frontend.

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.00% (+0.29 kb) | 0.00% (+0.29 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/kr/empty-spec-return-to-me/compare-report.html;tab=live_object) for a full comparison between the commits 905103c and 573a8bb or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

BolajiOlajide commented on migrations/frontend/squashed.sql:
```diff
@@ -821,7 +821,7 @@ CREATE TABLE batch_changes (
     updated_at timestamp with time zone DEFAULT now() NOT NULL,
     closed_at timestamp with time zone,
-    batch_spec_id bigint,
+    batch_spec_id bigint NOT NULL,
```
I think we'll need another migration to undo the changes.

courier-new replied:
This is actually the end state we want, I think. It used to be not null, https://github.com/sourcegraph/sourcegraph/pull/44840 made it null, this makes it not null again. Does that not match your expectations?

eseliger replied:
it does match mine, curious what Bolaji meant 🤔

BolajiOlajide replied:
I need to do a better job with notifications - I somehow missed this yesterday.

Yeah, I think you're both right - this is the expectation. I got confused for a bit there.

courier-new replied:
No worries, it's a lot of negatives haha. "This reverts the change to make it not not null so it's not null again." 🤪

eseliger commented on CHANGELOG.md:
```diff
@@ -45,7 +45,6 @@ All notable changes to Sourcegraph are documented in this file.
 - Code Insights: fixed an issue where filtering by a search context that included multiple repositories would exclude data. [#45574](https://github.com/sourcegraph/sourcegraph/pull/45574)
 - Ignore null JSON objects returned from GitHub API when listing public repositories. [#45969](https://github.com/sourcegraph/sourcegraph/pull/45969)
 - Fixed issue where emails that have never been verified before would be unable to receive resent verification emails. [#46185](https://github.com/sourcegraph/sourcegraph/pull/46185)
-- Batch Changes: Fixed resolution of the `currentSpec` field for draft batch changes on the GraphQL API. [#46237](https://github.com/sourcegraph/sourcegraph/pull/46237)
```
just a meta-question, if we fix something that was never released, would we want a changelog entry at all? Just came across my mind right now :D

courier-new replied:
At the time that I wrote this, I thought it _was_ released. 😂 But yeah probably not.

eseliger replied:
yeah no blame, was just genuinely curious what others think haha

eseliger reviewed (approved):
bye bye good database schema 😢 see you soon :)

courier-new commented:
2 other bugs I found and fixed out of this:

- The specs restored by the `up` migration weren't `created_from_raw=TRUE`, so their execution state wasn't being calculated appropriately. Both the original migration and new one have been updated.
- Since we made the change to exclude the empty specs from the batch specs list, the dynamic URL that a batch change on the list page redirects to on click has been broken. This was _also_ obscured by the "random spec" bug. 😂 I've fixed this [here](https://github.com/sourcegraph/sourcegraph/pull/46341/commits/4cf51a2dce79987146744661e8dd43a2678ba1d8).
